src/nupic/encoders/Scalar.cpp

```diff
@@ -0,0 +1,211 @@
+ * Numenta Platform for Intelligent Computing (NuPIC)
+ * Copyright (C) 2013, Numenta, Inc.  Unless you have an agreement
+ * with Numenta, Inc., for a separate license for this software code, the
+ * following terms and conditions apply:
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses.
+ *
+ * http://numenta.org/licenses/
+ * ---------------------------------------------------------------------
+ */
+
+/** @file
+ * Implementations of the ScalarEncoder and PeriodicScalarEncoder
+ */
+
+#include <cstring> // memset
+#include <cmath>
+#include <nupic/encoders/Scalar.hpp>
+#include <nupic/utils/Log.hpp>
+
+namespace nupic
+{
+  static Real64 extractScalar(const ArrayBase & input) {
+    switch (input.getType()) {
+    case NTA_BasicType_Byte:
+      return ((Byte*) input.getBuffer())[0];
+      break;
+    case NTA_BasicType_Int16:
+      return ((Int16*) input.getBuffer())[0];
+      break;
+    case NTA_BasicType_UInt16:
+      return ((UInt16*) input.getBuffer())[0];
+      break;
+    case NTA_BasicType_Int32:
+      return ((Int32*) input.getBuffer())[0];
+      break;
+    case NTA_BasicType_UInt32:
+      return ((UInt32*) input.getBuffer())[0];
+      break;
+    case NTA_BasicType_Int64:
+      return ((Int64*) input.getBuffer())[0];
+      break;
+    case NTA_BasicType_UInt64:
+      return ((UInt64*) input.getBuffer())[0];
+      break;
+    case NTA_BasicType_Real32:
+      return ((Real32*) input.getBuffer())[0];
+      break;
+    case NTA_BasicType_Real64:
+      return ((Real64*) input.getBuffer())[0];
+      break;
+    default:
+      NTA_THROW << "extractScalar: Unsupported type " << input.getType();
+    }
+  }
+
+  ScalarEncoder::ScalarEncoder(
+    int w, double minval, double maxval, int n, double radius,
+    double resolution, bool clipInput)
+    :w_(w),
+     minval_(minval),
+     maxval_(maxval),
+     clipInput_(clipInput)
+  {
+    if ((n != 0 && (radius != 0 || resolution != 0)) ||
+        (radius != 0 && (n != 0 || resolution != 0)) ||
+        (resolution != 0 && (n != 0 || radius != 0))) {
+      NTA_THROW <<
+        "Only one of n/radius/resolution can be specified for a ScalarEncoder.";
+    }
+
+    if (n != 0) {
+      n_ = n;
+      const int nBuckets = n - (w - 1);
+      // Distribute nBuckets points along the domain [minval, maxval], including
+      // the endpoints. The resolution is the distance between points.
+      resolution_ = (maxval - minval) / (nBuckets - 1);
```

Comment:
This will crash if `nBuckets` is `1`.  Not sure if we want nBuckets to be at least 1 or at least 2. Either way, put validation in for the acceptable range and document it.